prefect_gcp/cloud_storage.py

```diff
@@ -1140,3 +1141,77 @@ async def upload_from_folder(
         if num_uploaded == 0:
             self.logger.warning(f"No files were uploaded from {from_folder}.")
         return bucket_folder
+
+    @sync_compatible
+    async def upload_from_dataframe(
+        self,
+        df: pd.DataFrame,
+        to_path: str,
+        output_format: str = "parquet",
+        compression: Optional[str] = None,
+        **upload_kwargs: Dict[str, Any],
+    ) -> str:
+        """
+        Upload a pandas Dataframe to Google Cloud Storage as:
+            .csv, .csv.gz, .parquet, .parquet.snappy, .parquet.gz
+
+        Args:
+            df (pd.DataFrame): pandas Dataframe object
+            to_path (str): the actual full blob name (e.g.: /path/to/gcs/blob.csv)
+            output_format (str): Specify whether the output should be csv or parquet
+            compression (Optional[str], optional):
+                Specify the compression type for the output.
+                'csv' supports 'None' or 'gzip',
+                'parquet' supports 'None', 'snappy', 'gzip'.
+                Defaults to None.
+            **upload_kwargs: Additional keyword arguments to pass to
+                `Blob.upload_from_file`.
+
+        Returns:
+            The path that the folder was uploaded to.
+
+        Examples:
+            Upload local folder my_folder to the bucket's folder my_folder.
+            ```python
+            from prefect_gcp.cloud_storage import GcsBucket
+
+            gcs_bucket = GcsBucket.load("my-bucket")
+            gcs_bucket.upload_from_dataframe(
+                df=pandas_df, to_path="/path/to/gcs/blob.parquet",
+                output='parquet', compression='snappy'
+            )
+            ```
+        """
+        byte_buffer = BytesIO()
+        content_type: str = "application/octet-stream"
+
+        if not to_path.endswith(output_format):
+            to_path = f"{to_path}.{output_format}"
+
+        if output_format == "csv":
+            df.to_csv(path_or_buf=byte_buffer, compression=compression, index=False)
```

Comment:
Seems like pandas to csv can support dict compression as well (str + dict): https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html

Also, I wonder if we should handle the appending of file extensions for the users, or let them be smart about it (if they use compression, they should specify that in to_path) because I worry this function might get too complex to maintain

Reply:
Regarding the file extension thing, that's something I'm not so sure about either, because you see:

If they do something like `output=/path/to/gcs/blob.csv`, with `output='parquet'`, they'll literally end up with a blob on GCS named as: `/path/to/gcs/blob.csv.parquet`:

```python
gcs_bucket = GcsBucket.load("my-bucket")
gcs_bucket.upload_from_dataframe(
   df=pandas_df, 
   to_path="/path/to/gcs/blob.csv", 
   output='parquet', 
   compression='snappy'
)
```

<img width="844" alt="image" src="https://user-images.githubusercontent.com/901651/218227849-7027af86-cef4-493a-a628-1e82784ad9e8.png">

Reply:
Let me look into the dict for the `to_csv`, it might me worth a short doing something because of that `compress_level`. I didn't really bother since `.to_parquet( )` only supports a few set of strings

Reply:
Yes I meant we shouldn't handle it for them. If they set `to_path="blob.csv"` and output_format = "parquet" it'll show as `blob.csv`

Reply:
@ahuang11 I'm not gonna lie, I'm still scratching my head my head a bit since it may lead to confusion...

For instance, imagine this very scenario, you though:

```python
gcs_bucket.upload_from_dataframe(
   df=pandas_df, 
   to_path="/path/to/gcs/blob.csv", 
   output='parquet', 
   compression='gzip'
)
```

If we leave this unattended, the blob will be named `blob.csv`...

And if they eventually try to download that. to their machines, it will lead them into thinking they can open it on VSCode / Vi / Vim / Excel, and the second they do that, the app is gonna scream and they're gonna see a lot of garbage, leading them to think the file somehow got corrupted... :( 

Or if they try importing onto BigQuery, that will throw them errors, as they'll be trying to import it as a .csv, when, in fact, it's a .parquet """in disguize""". Does it make any sense?

Which is why I thought of enforcing the suffix, so we guarantee that regardless of the name, the extensions will always be right no matter what.

Reply:
Okay that makes sense to me. Perhaps we can highlight that in the docstring somewhere if we don't enforce it.

Reply:
I think I found an elegant way of fixing this, check this out (avoiding all of those IFs we had on the prototype):
(Doing a step-by-step so it gets easier to track, but we can chain some stuff in the real code)

```python
from pathlib import Path
...

to_path = "/some/path/to/gcs/blob.csv.gz"

path = Path(to_path).parent
### returns "/some/path/to/gcs/"

filename = Path(to_path).stem  
### returns "blob.csv"
# steam only chips one suffix at a time

filename = Path(filename_after_stem).with_suffix(".parquet.snappy")  
### returns: "blob.parquet.snappy"
# with_suffix adds or replaces a suffix (if exists) with the string you gave it

gcs_path = path.joinpath(filename)
### returns /some/path/to/gcs/blob.parquet.snappy
```

What do you think?